feat: KeyPair::from_seckey_byte_array

uncomputable opened the pull request:
Construct KeyPair directly from [u8; 32]. I can change the parameter to &[u8; 32] depending on the discussion in #780.

uncomputable commented:
Deprecation of a method is a candidate for the changelog.

uncomputable commented:
Fixed formatting

apoelstra commented:
Do we have serde regression tests here? I'm pretty nervous that the first commit changes the serialization format.

uncomputable commented:
This PR leaves the serialization format unchanged. The changes in `TupleVisitor{32,33}` merely affect how the byte array is passed to the parsing function.

Kixunil commented on src/serde_util.rs:
```diff
@@ -106,7 +106,7 @@ macro_rules! impl_tuple_visitor {
                         return Err(de::Error::invalid_length(i, &self));
                     }
                 }
```
Oh, wow, it looks like this code allows for longer bytes (just not shorter).

Kixunil reviewed (approved):
ACK e7eea32ffaf36172a03063b634a435ec937b7d21

I'm convinced this doesn't change serialization and whatever problems there are were there already.

apoelstra commented on src/key.rs:
```diff
@@ -1597,7 +1597,7 @@ impl<'de> serde::Deserialize<'de> for XOnlyPublicKey {
-                XOnlyPublicKey::from_slice,
+                |bytes| XOnlyPublicKey::from_byte_array(&bytes),
             );
             d.deserialize_tuple(constants::SCHNORR_PUBLIC_KEY_SIZE, visitor)
```
I think that bytes would add a length prefix in some serializations. Or possibly this was an artifact of the early days when trying to get the `serde_test` serializer to round-trip, which had tons of random-seeming problems with `Bytes` and `ByteSlice`.

In any case, if this PR doesn't change it then it's out of scope. We can revisit this separately. (Or not, if we decide that the existing behavior is acceptable.)

Kixunil replied:
> I think that bytes would add a length prefix in some serializations.

Oh, yes, I think so. Good point.

uncomputable commented:
I updated the other key constructors to take owned array values. I updated the PR title and added an entry to the changelog for a breaking change.

@Kixunil I didn't see your approval until now; sorry.

apoelstra commented:
WASM we can ignore. We should probably just drop the CI job entirely until they fix my 18-month-old issue [about wasm-pack not working at all in CI setups](https://github.com/rustwasm/wasm-pack/issues/1316#issuecomment-2307266309).

apoelstra commented:
>Let's see if I can fix WASM in a follow-up PR.

Thanks! I would appreciate that, though bear in mind that I will never test it locally until they fix `wasm-pack` so that it's possible to run it without it attempting to download and execute random things.

Kixunil reviewed (approved):
ACK 59cf7bd52c208a83d26ea16f2a27be2f8787b5d5

Just FYI, deprecation is *not* a breaking change (yet), only removal is. Still, the changelog note is warranted.

apoelstra reviewed (approved):
ACK 59cf7bd52c208a83d26ea16f2a27be2f8787b5d5; successfully ran local tests; nice! the old function signatures were super weird. and the docs were wrong lol

tcharding reviewed (approved):
ACK 59cf7bd52c208a83d26ea16f2a27be2f8787b5d5
